include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+      {
+        return (f(center + step) - 2.0*f(center) + f(center - step))/(step*step);
+      }
+
+
+
+      template <int structdim, typename F>
+      inline
+      auto
+      cross_stencil
+      (const CrossDerivative                                        cross_derivative,
+       const Tensor<1, GeometryInfo<structdim>::vertices_per_cell> &center,
+       const double                                                 step,
+       const F                                                     &f)
+      -> decltype(f(center) - f(center))
+      {
+        Tensor<1, GeometryInfo<structdim>::vertices_per_cell> simplex_vector;
+        simplex_vector[cross_derivative.direction_0] = 0.5*step;
+        simplex_vector[cross_derivative.direction_1] = -0.5*step;
+        return (f(center + simplex_vector) - f(center - simplex_vector))/step;
+      }
+
+
+
+      template <int spacedim, int structdim, typename F>
+      inline
+      double
+      gradient_entry
+      (const unsigned int                                           row_n,
+       const unsigned int                                           dependent_direction,
+       const Point<spacedim>                                       &p0,
+       const Tensor<1, GeometryInfo<structdim>::vertices_per_cell> &center,
+       const double                                                 step,
+       const F                                                     &f)
+      {
+        Assert(row_n < GeometryInfo<structdim>::vertices_per_cell &&
+               dependent_direction < GeometryInfo<structdim>::vertices_per_cell,
+               ExcMessage("This function assumes that last weight is a dependent "
+                          "variable (and hence we cannot take its derivative "
+                          "directly)."));
+        Assert(row_n != dependent_direction,
+               ExcMessage("We cannot differentiate with respect to the variable "
+                          "that is assumed to be dependent."));
+
+        const Point<spacedim> manifold_point = f(center);
+        const Tensor<1, spacedim> stencil_value = cross_stencil<structdim>
+                                                  ({row_n, dependent_direction},
+                                                   center,
+                                                   step,
+                                                   f);
+        double entry = 0.0;
+        for (unsigned int dim_n = 0; dim_n < spacedim; ++dim_n)
+          entry += -2.0*(p0[dim_n] - manifold_point[dim_n])*stencil_value[dim_n];
+        return entry;
+      }
+
+      /**
+       * Project onto a flat object. This is more accurate than the general
+       * algorithm in project_to_object.
+       */
+      template <typename Iterator, int spacedim, int structdim>
+      Point<spacedim>
+      project_to_flat_object (const Iterator        &object,
+                              const Point<spacedim> &trial_point)
```

Comment:
why is this function called `project_to_FLAT_object`? It just minimizes a distance, but the object can be anything that's described by a Q1 mapping, right?

Reply:
Yes. This is a copy-and-paste of the old `StraightBoundary` algorithm.

A better name would be 'project_to_d_linear`.

Reply:
Yes, I agree. Go with it.